lib/events/search_limiter.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package events
+
+import (
+	"math"
+	"time"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/time/rate"
+
+	"github.com/gravitational/teleport/api/types"
+	apievents "github.com/gravitational/teleport/api/types/events"
+)
+
+// SearchEventsLimiter allows to wrap any AuditLogger with rate limit on
+// search events endpoints.
+// It can be used when AWS qoutas are tight.
+// Note it share limiter for both SearchEvents and SearchSessionEvents.
+type SearchEventsLimiter struct {
+	limiter *rate.Limiter
+	AuditLogger
+}
+
+type SearchEventsLimiterConfig struct {
+	LimiterRate  float64
+	LimiterBurst int
+	AuditLogger  AuditLogger
+}
+
+func (cfg *SearchEventsLimiterConfig) CheckAndSetDefaults() error {
+	if cfg.AuditLogger == nil {
+		return trace.BadParameter("empty auditLogger")
+	}
+	if cfg.LimiterBurst <= 0 {
+		return trace.BadParameter("limiterBurst cannot be less or equal to 0")
+	}
+	if math.Abs(cfg.LimiterRate) < 1e-9 {
```

Comment:
sorry for missing docs for that config. `LimiterRate` defines how many tokens and added per seconds. 
I am checking it with `1e-9` just to have very small delta when comparing float against 0. On one hand it should be equal to 0 because if it's not present in url params, it will be default, so 0. @rosstimothy do you think I should just better document it? 

@smallinsky i don't think you can define rps using uint, someone may want to set it 0.5 for example. And defining it in miliseconds will be even more confusing. 

One thing that we can do is use helper fn from `rate` package [Every](https://pkg.go.dev/golang.org/x/time/rate#Every). You define interval (time.Duration) at which token is filled. Do you think it will be easier to understand?

Reply:
0.5 RPS or 10.1 RPS values as rate limit should be quite rate but I don't have strong about it so please free to ignore my comment.

> I don't thing that we can do is use helper fn from rate package [Every](https://pkg.go.dev/golang.org/x/time/rate#Every). You define interval (time.Duration) at which token is filled. Do you think it will be easier to understand?

I think that current behaviour with 1s time unit is fine and readable. It should be quite easy to extend in the future but if we don't have case for Athena we should skip it for now and keep it simple.